specs/NestedFrame.md

```diff
@@ -0,0 +1,215 @@
+CoreWebView2Frame.FrameCreated API
+===
+
+# Background
+At present, WebView2 enables developers to track only first-level
+iframes, which are the direct child iframes of the main frame.
+However, we see that WebView2 customers want to manage nested
+iframes, such as recording the navigation history for a second
+level iframe. To address this, we will introduce the
+`CoreWebView2Frame.FrameCreated` API. This new API will allow
+developers to subscribe to the nested iframe creation event,
+giving them access to all properties, methods, and events of
+[CoreWebView2Frame](https://learn.microsoft.com/dotnet/api/microsoft.web.webview2.core.corewebview2frame)
+for the nested iframe.
+
+To prevent unnecessary performance implication, WebView2 does
+not track any nested iframes by default. It only tracks a nested

```

Comment:
This seems to [violate Windows Runtime guidelines](https://www.osgwiki.com/wiki/WinRT_API_Design_Guidelines#Event_Design), because it means that you don't get a CoreWebView2.FrameNavigationStarting unless you subscribe to the FrameCreated event (even if the event handler does nothing).

> ❌ DO NOT alter behavior based on whether an event handler is registered or the number of registered event handlers. In other words, registering an event handler that does nothing should have no effect on behavior.
> 
> An implementation is permitted to perform optimizations based on whether an event handler is registered. The most common case of this is bypassing an expensive operation if there is no way for the app to observe any of the operation’s side effects.

In this case, the app can observe the side effects (through the FrameNavigationStarting event).

We could have a new Boolean property on the frame `ShouldRaiseNavigationEvents` which defaults to true for first-level frames and false for deeper frames. (But see the componentization problem later.)

Reply:
I need to clarify the documentation here. The `CoreWebView2.FrameNavigationStarting` event will still be triggered for all child iframes, regardless of whether we subscribe to the `CoreWebView2Frame.FrameCreated` event or not. In this context, we are referring to the APIs in [CoreWebView2Frame](https://learn.microsoft.com/en-us/microsoft-edge/webview2/reference/winrt/microsoft_web_webview2_core/corewebview2frame?view=webview2-winrt-1.0.2903.40). We think that event handler matters because if developers do not subscribe to it, there is no need to expose any CoreWebView2Frame APIs

Reply:
Please clarify that this is an implementation note. This is only about an internal perf optimization - no observable behavior changes described here.